src/lib/core/fiber.c

```diff
@@ -1498,6 +1516,35 @@ cord_create(struct cord *cord, const char *name)
 #endif
 }
 
+/**
+ * Delete the latest garbage fiber which couldn't be deleted somewhy before. Can
+ * safely rely on the fiber being not the current one. Because if it was added
+ * here before, it means some previous fiber put itself here, then died
+ * immediately afterwards for good, and gave control to another fiber. It
+ * couldn't be scheduled again.
+ */
+void
+cord_collect_garbage(struct cord *cord)
+{
+	struct fiber *garbage = cord->garbage;
+	if (garbage == NULL)
+		return;
+	assert(cord->fiber != garbage);
+	fiber_delete(cord, garbage);
+	cord->garbage = NULL;
+}
+
+static void
+cord_add_garbage(struct cord *cord, struct fiber *f)
+{
+	cord_collect_garbage(cord);
+	assert(cord->garbage == NULL);
```

Comment:
If as *'it'* you mean the fiber `f`, then we can't delete the current fiber right away. Its stack is still needed and we need a valid `cord->fiber` in order to yield to another fiber. About the stack see the comment in `cord.garbage` member:
```C
	/**
	 * Latest dead fiber which couldn't be reused and waits for its
	 * deletion. A fiber can't be reused if it is somehow non-standard. For
	 * instance, has a special stack.
	 * A fiber can't be deleted if it is the current fiber - can't delete
	 * own stack safely. Then it schedules own deletion for later. The
	 * approach is very similar to pthread stacks deletion - pthread can't
	 * delete own stack, so they are saved and deleted later by a newer
	 * pthread or by some other dying pthread. Same here with fibers.
	 */
	struct fiber *garbage;
```
In particular: *A fiber can't be deleted if it is the current fiber - can't delete own stack safely*. You won't even be able to return from `fiber_recycle()` if you free the stack you are standing on.

Reply:
No, I meant the `cord->garbage` that you asserting to be `NULL`. What if it's not `NULL` - it can be safely deleted, since it was put there before, supposedly by a different fiber.

Reply:
It can't be not `NULL`. I called `cord_collect_garbage` on the previous line.

Reply:
sorry, my comment was related to an outdated patch.
looks good!

Reply:
You need to 'approve' the patch via Github. Otherwise I can't merge it.